### ISSUE TITLE
FSO Locking up when receiving user input during load states under Windows.

### ISSUE THREAD

**Issue by JadedDragoon** [scp-fs2open/fs2open.github.com#3622]
To be clear... I've not seen any cases of this happening when a loading bar is displayed on the screen. This usually occurs when FSO is busy enough that the windows mouse cursor reappears and shows the "busy" animation.

Examples include:

- between pressing "commit" and actually seeing the in-flight hud on screen
- during initial start of FSO before the intro movie plays
- after the intro movies plays but before the player select screen appears

I've been having this issue at least since 21.0, possibly before. _any_ user input can cause this while the fso windows has focus... including user input not intended for FSO specifically (such as pausing or unpausing music with keyboard keys or usig the volume roller on my keyboard to raise or lower the volume). Even a single mouse click can result in the issue.

Once the issue has been triggered Windows turned the FSO window grey and displays the usual windows "this program is not responding... wait/close" message. Choosing to wait simply leaves FSO locked up. Ultimately, the only way forward becomes to close FSO.

To [quote](https://discord.com/channels/223511295431933953/223511391531827200/880093005518897154) Lafiel from Discord #SCP :

> So, what's happening here is an operation system issue.
If you attach it to a debugger, this doesn't happen.
If it's not attached to a debugger, windows sees that the app isn't taking in user commands, and it's then windows that decides that this app will be halted.
You can see that if you cause it to freeze up at first, with the slightly white-ish screen, it still continues to load as normal, but once windows decides "This isn't good", and shows you the "Close Program" Prompt, it'll halt FSO, and seemingly not properly unhalt when you click wait.
The proper solution would be to accept and then disregard inputs during load states, but I have no clue how to make that happen in SDL and the FSO input system

**Comment by Goober5000** [comment]
I assume that `os_poll()` needs to be called on a regular basis, and that something is causing too much time to pass between calls.

**Comment by asarium** [comment]
For the game loading screen, we have something in place that calls `os_poll()` whenever we update the progress indicator but those events are buffered instead of being delivered to their respective systems.

If a certain action takes long enough then the same thing might be necessary there. However, even if we run into that unresponsive state at some point we should recover once we get into some code that regularly calls `os_poll`. I can't explain why that does not happen though...

**Comment by BMagnu** [comment]
In my testing, windows explicitly halts the FSO process OS-side once it determines it's unresponsive.
This'd explain why it doesn't recover.

**Comment by asarium** [comment]
What does "halt" mean here? If Windows just suspends the process then this is a Windows bug and not something FSO can really fix reliably.

**Comment by BMagnu** [comment]
Yes, windows seems to suspend the process as far as I can tell. But FSO can fix it by not having too large gaps inbetween os_poll, as far as I can tell.
And since os_poll is probably not too expensive, there should be little harm in calling it more often, especially for large levels.

**Comment by asarium** [comment]
os_poll needs to be used with care though. It causes all sorts of related actions to take place so if it is not being called at the end of a frame `os_ignore_events` should be used instead which buffers the events and replays them the next time we actually call `os_poll`.

Also, if something outside of level loading takes so much time, maybe that should be looked at in more detail and handled better.

**Comment by JadedDragoon** [comment]
> For the game loading screen, we have something in place that calls `os_poll()` whenever we update the progress indicator but those events are buffered instead of being delivered to their respective systems.
> 
> If a certain action takes long enough then the same thing might be necessary there. However, even if we run into that unresponsive state at some point we should recover once we get into some code that regularly calls `os_poll`. I can't explain why that does not happen though...

I can say that I only recently managed to get this issue to occur on the loading screen. The issue is much more common in the example cases I listed in the OP. So apparently the mitigation steps taken for the loading screen _are_ helping.

Also, when talking about Windows "halting FSO" we are talking about this:
![image](https://user-images.githubusercontent.com/1294429/132106038-2a3d8ac1-b4c3-44f5-a36d-61ecec9844a6.png)

With most other applications that stop responding and windows displays the "is not responding" dialog... clicking "wait" allows the application to eventually continue unless the application is completely locked up. With FSO it does not... once this dialog appears FSO _becomes_ completely locked up. I am not a particularly experienced software engineer... but I feel like that is where the focus here needs to be. Not on keeping the "is not responding" dialog for ever appearing under any circumstances... but on figuring out why FSO in particular can never recover once the above dialog is displayed. As this issue happens with FSO and not the hundreds of thousands of other games and applications that get halted by windows for various reasons... I'd argue this is still an issue with FSO.

**Comment by BMagnu** [comment]
If you have the task manager open while this happens, you can see that teh status field for FSO will display something that translates to halted or suspended as soon as this dialog pops.
Other applications don't do this when this happens, and I believe it's what is preventing FSO from recovering.
Ideally though, it'd os_poll or os_ignore_events often enough to never even get to this point anyways...

**Comment by z64555** [comment]
I'm chucking this to 21.6 and marking it as a low priority, since we have a work around for it (by simply not jamming down any inputs).  If it becomes more frequent of a problem we can elevate it to high
